### PR TITLE
Extend startup notification timeout as each initialization milestone is attained

### PR DESCRIPTION
**Pull request by basil**
A potential problem with the startup notification support introduced in #6228 is very slow Jenkins installations. By default `systemd` waits up to 90 seconds for a service to start up before the start timeout is reached. I have seen some slow Jenkins installations that take longer than that. This PR introduces a new `Lifecycle` API for pushing back the timeout and adjusts the initialization logic to push the timeout back by 15 seconds as each initialization milestone is attained `STARTED`, `PLUGINS_LISTED`, `PLUGINS_PREPARED`, `PLUGINS_STARTED`, `EXTENSIONS_AUGMENTED`, `SYSTEM_CONFIG_LOADED`, `SYSTEM_CONFIG_ADAPTED`, `JOB_LOADED`, `JOB_CONFIG_ADAPTED`, and `COMPLETED`, so potentially increasing the timeout by another 2.5 minutes, as long as Jenkins is actually making progress in starting up. This should make it far less likely that people hit startup timeouts when using `systemd` with `Type=notify`, unless people have a huge amount of jobs that take many minutes to load. Even in that situation, we should be able to consume this new API from e.g. https://github.com/jenkinsci/cloudbees-folder-plugin/blob/d63f3c0efe1b951aba3278f671cd6dea192908ec/src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolder.java#L563-L581 to push back the timeout every 15 seconds or so as long as Jenkins continues to make progress in loading jobs.

To test this, I set the overall service timeout to an absurdly low value (2 seconds) and verified that the service still started up (in about 10 seconds), confirming that the timeout extension system was working.

### Proposed changelog entries

* Extend startup notification timeout as each initialization milestone is attained.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
